real_time/arduino_src/lib_avr/encoder/encoder.cpp

```diff
@@ -0,0 +1,170 @@
+#include "encoder.h"
+
+
+Encoder::Encoder() {
+}
+
+
+uint8_t Encoder::Init(volatile uint8_t *pin_a_reg,
+                      uint8_t pin_a_num) {
+    // save configuration info
+    pin_a_reg_ = pin_a_reg;
+    pin_a_num_ = pin_a_num;
+
+    // reset state
+    ticks_ = 0;
+    errors_ = 0;
+
+    return errors_;
+}
+
+
+uint8_t Encoder::InitQuad(volatile uint8_t *pin_a_reg,
+                          uint8_t pin_a_num,
+                          volatile uint8_t *pin_b_reg,
+                          uint8_t pin_b_num) {
+    // save configuration info
+    pin_a_reg_ = pin_a_reg;
+    pin_a_num_ = pin_a_num;
+    pin_b_reg_ = pin_b_reg;
+    pin_b_num_ = pin_b_num;
+    // reset state
+    ticks_ = 0;
+    bool pin_a = (*pin_a_reg_) & _BV(pin_a_num_);
+    bool pin_b = (*pin_b_reg_) & _BV(pin_b_num_);
+    pin_state_last_ = ((uint8_t)pin_a << 1) | ((uint8_t)pin_b);
+    errors_ = 0;
+    // set required pins as inputs
+    // TODO
```

Comment:
They're not set right now because we don't pass enough information through in the Init() call. This should be cleaned up once we have an interrupt setup class.